Index: native_client/src/trusted/service_runtime/arch/x86_64/nacl_syscall_64.S
===================================================================
--- native_client/src/trusted/service_runtime/arch/x86_64/nacl_syscall_64.S	(revision 3679)
+++ native_client/src/trusted/service_runtime/arch/x86_64/nacl_syscall_64.S	(working copy)
@@ -62,7 +62,14 @@
          */
 #if NACL_LINUX
         /* Linux uses TLS. */
+#ifdef __PIC__
+        leaq nacl_thread_index@TLSGD(%rip), %rdi
+        rex64
+        call __tls_get_addr@PLT
+        movl (%rax), %eax
+#else
         movl    %fs:nacl_thread_index@TPOFF, %eax
+#endif
 #elif NACL_OSX
         /*
          * This assumes a particular implementation of OS X's
